dace/sdfg/state.py

```diff
@@ -756,37 +762,59 @@ def _read_and_write_sets(self) -> Tuple[Dict[AnyStr, List[Subset]], Dict[AnyStr,
-                        if e.data.is_empty():
+                if not isinstance(n, nd.AccessNode):
+                    continue
+                ac_desc = n.desc(self.sdfg)
+                ac_size = ac_desc.total_size
+                in_edges = [in_edge for in_edge in sg.in_edges(n) if not in_edge.data.is_empty()]
+                out_edges = [out_edge for out_edge in sg.out_edges(n) if not out_edge.data.is_empty()]
+
+                # In some conditions subsets can be `None`, we will now clean them.
+                in_subsets = dict()
+                for in_edge in in_edges:
+                    assert in_edge.data.dst_subset is not None or (in_edge.data.num_elements() == ac_size)
+                    in_subsets[in_edge] = (
+                            sbs.Range.from_array(ac_desc)
+                            if in_edge.data.dst_subset is None
+                            else in_edge.data.dst_subset
+                        )
+                out_subsets = dict()
+                for out_edge in out_edges:
+                    assert (out_edge.data.src_subset is not None) or (out_edge.data.num_elements() == ac_size)
+                    out_subsets[out_edge] = (
+                        sbs.Range.from_array(ac_desc)
+                        if out_edge.data.src_subset is None
+                        else out_edge.data.src_subset
+                    )
+
+                # Filter out memlets which go out but the same data is written to the AccessNode by another memlet
+                for out_edge in list(out_edges):
+                    for in_edge in in_edges:
+                        if out_edge.data.data != in_edge.data.data:
+                            # NOTE: This check does not make any sense, and is in my view wrong.
+                            #   As it will filter out some accesses but not all, which one solely
+                            #   depends on how the memelts were created.
+                            #   See also [issue #1643](https://github.com/spcl/dace/issues/1643).
                             continue
-                        rs[n.data].append(e.data.subset)
+                        if in_subsets[in_edge].covers(out_subsets[out_edge]):
+                            out_edges.remove(out_edge)
+                            break
+
+                if in_edges:
+                    ws[n.data].extend(in_subsets.values())
+                if out_edges:
+                    rs[n.data].extend(out_subsets[out_edge] for out_edge in out_edges)
+
             # Union all subgraphs, so an array that was excluded from the read
             # set because it was written first is still included if it is read
             # in another subgraph
             for data, accesses in rs.items():
                 read_set[data] += accesses
             for data, accesses in ws.items():
                 write_set[data] += accesses
-        return read_set, write_set
+        return copy.deepcopy((read_set, write_set))
```

Comment:
Why is it necessary to make a copy here?

Reply:
Because, the subsets are still linked to the Memlets, so if you modify them then you change them at the Memlets.
This might be useful but since it is not possible to determine which subset belongs to which memlet, it does not make sense to maintain this link.
You could potentially copy them above, because some of them are constructed on demand anyway, however:
- The on demand construction is the minority case.
- Doing it here allows to do it in one big sweep.